File: scripts/backend-node/qrService.js
// Create Dynamic QR — POST /paymentservices/qr/create
// Doc: https://www.paytmpayments.com/docs/api/create-qr-code-api
// Defaults & gotchas baked in:
//   - posId is REQUIRED (omitting it returns 400)
//   - amount must be a STRING with two decimals
//   - head requires clientId + version + signature
//   - Response `image` is RAW base64 — we prepend `data:image/png;base64,` here
//     so the frontend can drop it straight into <img src>.
import crypto from "node:crypto";
import PaytmChecksum from "paytmchecksum";
import { getPaytmConfig } from "./paytmConfig.js";

function normalizeAmount(amount) {
  const n = Number(String(amount ?? "").trim());
  if (!Number.isFinite(n) || n <= 0) return "1.00";
  return n.toFixed(2);
}

export async function createDynamicQr({
  amount,
  posId = "POS001",
  displayName,
  expiryDate,
  imageRequired = true,
  orderId: callerOrderId,
}) {
  const cfg = getPaytmConfig();
  if (!cfg.mid) throw asError("MISSING_MID", "Missing PAYTM_MID");
  if (!cfg.merchantKey) throw asError("MISSING_MERCHANT_KEY", "Missing PAYTM_MERCHANT_KEY");
  if (!posId || !String(posId).trim()) {
    throw asError("MISSING_POS_ID", "posId is required for QR creation (Paytm returns 400 without it)");
  }

  const orderId = callerOrderId?.trim()
    || `QR_${crypto.randomBytes(10).toString("hex").toUpperCase()}`;

  const body = {
    mid: cfg.mid,
    orderId,
    amount: normalizeAmount(amount),
    businessType: "UPI_QR_CODE",
    posId: String(posId).trim(),
    imageRequired: !!imageRequired,
    ...(displayName?.trim() ? { displayName: displayName.trim().slice(0, 30) } : {}),
    ...(expiryDate?.trim() ? { expiryDate: expiryDate.trim() } : {}),
  };

  const signature = await PaytmChecksum.generateSignature(JSON.stringify(body), cfg.merchantKey);
  const head = {
    clientId: cfg.clientId,
    version: "v1",
    signature,
  };

  const r = await fetch(cfg.qrCreateUrl, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ head, body }),
  });
  const text = await r.text();
  if (!r.ok) throw upstream("QR_HTTP_ERROR", `qr/create HTTP ${r.status}`, orderId, text);

  const json = JSON.parse(text);
  const status = json?.body?.resultInfo?.resultStatus;
  if (status && status !== "SUCCESS" && status !== "S") {
    throw upstream("QR_FAILED", json?.body?.resultInfo?.resultMsg || "qr/create failed", orderId, text);
  }

  // Paytm returns `image` as RAW base64 (no data-URI prefix). Inject the prefix
  // server-side so the frontend can drop it into <img src> without surprises.
  const rawImage = json?.body?.image;
  const image = rawImage ? `data:image/png;base64,${rawImage}` : null;

  return {
    orderId,
    qrCodeId: json?.body?.qrCodeId,
    qrData: json?.body?.qrData,
    image,
    mid: cfg.mid,
  };
}

function asError(code, message) {
  const e = new Error(message);
  e.code = code;
  return e;
}

function upstream(code, message, orderId, raw) {
  const e = new Error(message);
  e.code = code;
  e.httpStatus = 502;
  e.orderId = orderId;
  try {
    const info = JSON.parse(raw)?.body?.resultInfo;
    if (info) e.paytm = { resultStatus: info.resultStatus, resultCode: info.resultCode, resultMsg: info.resultMsg };
  } catch {}
  return e;
}

File: scripts/backend-python/qr_service.py
"""Create Dynamic QR — POST /paymentservices/qr/create.

Doc: https://www.paytmpayments.com/docs/api/create-qr-code-api
Defaults & gotchas baked in:
- posId is REQUIRED (omitting it returns 400)
- amount must be a STRING with two decimals
- head requires clientId + version + signature
- Response `image` is RAW base64 — we prepend `data:image/png;base64,` here so
  the frontend can drop it straight into <img src>.
"""
import json
import secrets
from typing import Any, Optional

import requests
from paytmchecksum import PaytmChecksum

from paytm_config import get_paytm_config
from paytm_service import PaytmError


def _normalize_amount(amount: Any) -> str:
    try:
        n = float(str(amount).strip())
    except (TypeError, ValueError):
        return "1.00"
    if n <= 0:
        return "1.00"
    return f"{n:.2f}"


def create_dynamic_qr(
    *,
    amount: Any = "1.00",
    pos_id: str = "POS001",
    display_name: Optional[str] = None,
    expiry_date: Optional[str] = None,
    image_required: bool = True,
    order_id: Optional[str] = None,
) -> dict:
    cfg = get_paytm_config()
    if not cfg["mid"]:
        raise PaytmError("MISSING_MID", "Missing PAYTM_MID")
    if not cfg["merchant_key"]:
        raise PaytmError("MISSING_MERCHANT_KEY", "Missing PAYTM_MERCHANT_KEY")
    if not pos_id or not str(pos_id).strip():
        raise PaytmError("MISSING_POS_ID",
                         "posId is required for QR creation (Paytm returns 400 without it)")

    order_id = (order_id or "").strip() or ("QR_" + secrets.token_hex(10).upper())

    body: dict = {
        "mid": cfg["mid"],
        "orderId": order_id,
        "amount": _normalize_amount(amount),
        "businessType": "UPI_QR_CODE",
        "posId": str(pos_id).strip(),
        "imageRequired": bool(image_required),
    }
    if display_name and display_name.strip():
        body["displayName"] = display_name.strip()[:30]
    if expiry_date and expiry_date.strip():
        body["expiryDate"] = expiry_date.strip()

    signature = PaytmChecksum.generateSignature(json.dumps(body), cfg["merchant_key"])
    head = {
        "clientId": cfg["client_id"],
        "version": "v1",
        "signature": signature,
    }

    r = requests.post(cfg["qr_create_url"], json={"head": head, "body": body},
                      headers={"Content-Type": "application/json"}, timeout=15)
    text = r.text
    if not r.ok:
        raise _upstream("QR_HTTP_ERROR", f"qr/create HTTP {r.status_code}", order_id, text)

    data = json.loads(text)
    info = (data.get("body") or {}).get("resultInfo") or {}
    status = info.get("resultStatus")
    if status and status not in ("SUCCESS", "S"):
        raise _upstream("QR_FAILED", info.get("resultMsg") or "qr/create failed", order_id, text)

    body_resp = data.get("body") or {}
    raw_image = body_resp.get("image")
    image = f"data:image/png;base64,{raw_image}" if raw_image else None

    return {
        "orderId": order_id,
        "qrCodeId": body_resp.get("qrCodeId"),
        "qrData": body_resp.get("qrData"),
        "image": image,
        "mid": cfg["mid"],
    }


def _upstream(code: str, message: str, order_id: str, raw: str) -> PaytmError:
    paytm: dict = {}
    try:
        info = ((json.loads(raw) or {}).get("body") or {}).get("resultInfo") or {}
        for k in ("resultStatus", "resultCode", "resultMsg"):
            if info.get(k):
                paytm[k] = info[k]
    except Exception:
        pass
    return PaytmError(code, message, http_status=502, order_id=order_id, paytm=paytm or None)
